File: java-advent-of-code/src/main/java/com/mycompany/advent_of_code_java/Advent_of_code_2025_002_002.java
package com.mycompany.advent_of_code_java;

import module java.base;

public class Advent_of_code_2025_002_002 {

    public static void main(String[] args) {
        String inputPath = (args.length >= 1)
                ? args[0]
                : "E:\\documents-container\\advent-of-code\\java-advent-of-code\\inputs\\2025-002\\input.txt";

        try {
            BigInteger ans = solve(Path.of(inputPath));
            System.out.println(ans);
        } catch (IOException e) {
            System.err.println("Failed to read input: " + inputPath);
            e.printStackTrace(System.err);
            System.exit(1);
        }
    }

    public static BigInteger solve(Path inputFile) throws IOException {
        String line = Files.readString(inputFile).trim();
        if (line.isEmpty()) return BigInteger.ZERO;

        String[] parts = line.split(",");
        long[][] ranges = new long[parts.length][2];

        long globalMax = 0;
        for (int i = 0; i < parts.length; i++) {
            String token = parts[i].trim();
            if (token.isEmpty()) continue;

            int dash = token.indexOf('-');
            if (dash < 0) throw new IllegalArgumentException("Bad range token: " + token);

            long a = Long.parseLong(token.substring(0, dash).trim());
            long b = Long.parseLong(token.substring(dash + 1).trim());
            if (a > b) { long t = a; a = b; b = t; }

            ranges[i][0] = a;
            ranges[i][1] = b;
            if (b > globalMax) globalMax = b;
        }

        // Generate all invalid IDs <= globalMax (unique)
        long[] invalid = generateAllRepeatedNumbersUpTo(globalMax);

        // Prefix sums for O(log N) range sums
        BigInteger[] pref = new BigInteger[invalid.length + 1];
        pref[0] = BigInteger.ZERO;
        for (int i = 0; i < invalid.length; i++) {
            pref[i + 1] = pref[i].add(BigInteger.valueOf(invalid[i]));
        }

        BigInteger total = BigInteger.ZERO;
        for (long[] r : ranges) {
            long A = r[0], B = r[1];
            int left = lowerBound(invalid, A);
            int right = upperBound(invalid, B);
            if (left < right) {
                total = total.add(pref[right].subtract(pref[left]));
            }
        }

        return total;
    }

    /**
     * Generates all numbers <= max that can be written as some digit block repeated >=2 times,
     * with no leading zeros (so the block must be a k-digit number with first digit != 0).
     *
     * Returns sorted unique array.
     */
    private static long[] generateAllRepeatedNumbersUpTo(long max) {
        if (max < 11) { // smallest repetition with no leading zeros is 11
            return new long[0];
        }

        int maxDigits = digits(max);

        // Precompute powers of 10 up to maxDigits
        long[] pow10 = new long[maxDigits + 1];
        pow10[0] = 1L;
        for (int i = 1; i <= maxDigits; i++) {
            pow10[i] = pow10[i - 1] * 10L;
        }

        LongList list = new LongList(120_000);

        // Total length n, block length k, repeat count t = n/k >= 2
        for (int n = 2; n <= maxDigits; n++) {
            for (int k = 1; k < n; k++) {
                if (n % k != 0) continue;
                int t = n / k;
                if (t < 2) continue;

                long blockPow = pow10[k];
                long startX = pow10[k - 1];     // smallest k-digit (no leading zero)
                long endX = pow10[k] - 1;       // largest k-digit

                for (long x = startX; x <= endX; x++) {
                    long v = 0L;

                    // build v = repeat x, t times: v = ((x)*10^k + x)*10^k + x ... t times
                    for (int rep = 0; rep < t; rep++) {
                        v = v * blockPow + x;
                        // v cannot overflow for our digit sizes (<= 10 digits here), but keep it simple.
                    }

                    if (v > max) {
                        // For fixed (n,k,t), v increases with x, so we can break early.
                        break;
                    }

                    list.add(v);
                }
            }
        }

        long[] arr = list.toArray();
        Arrays.sort(arr);

        // Deduplicate (e.g., 1111 can arise from "1 repeated 4" and "11 repeated 2")
        int m = 0;
        for (int i = 0; i < arr.length; i++) {
            if (i == 0 || arr[i] != arr[i - 1]) {
                arr[m++] = arr[i];
            }
        }
        return Arrays.copyOf(arr, m);
    }

    private static int digits(long x) {
        int d = 0;
        long t = x;
        while (t > 0) {
            d++;
            t /= 10;
        }
        return Math.max(d, 1);
    }

    // Standard lower_bound: first index i with a[i] >= key
    private static int lowerBound(long[] a, long key) {
        int lo = 0, hi = a.length;
        while (lo < hi) {
            int mid = (lo + hi) >>> 1;
            if (a[mid] < key) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    }

    // Standard upper_bound: first index i with a[i] > key
    private static int upperBound(long[] a, long key) {
        int lo = 0, hi = a.length;
        while (lo < hi) {
            int mid = (lo + hi) >>> 1;
            if (a[mid] <= key) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    }

    // Simple dynamic long list (avoids boxing)
    private static final class LongList {
        private long[] data;
        private int size;

        LongList(int initialCapacity) {
            data = new long[Math.max(16, initialCapacity)];
            size = 0;
        }

        void add(long v) {
            if (size == data.length) {
                data = Arrays.copyOf(data, data.length * 2);
            }
            data[size++] = v;
        }

        long[] toArray() {
            return Arrays.copyOf(data, size);
        }
    }
}

File: python/Advent_of_code_2025_002_002.py
import sys
from pathlib import Path
import bisect

def get_digits_count(x: int) -> int:
    if x == 0:
        return 1
    return len(str(x))

def generate_all_repeated_numbers_up_to(max_val: int) -> list[int]:
    """
    Generates all numbers <= max that are formed by repeating a block of digits.
    Example: 1212 (block 12 repeated twice), 111 (block 1 repeated thrice).
    """
    if max_val < 11:
        return []

    max_digits = get_digits_count(max_val)
    # Precompute powers of 10
    pow10 = [10**i for i in range(max_digits + 1)]
    
    results = set() # Use a set to handle deduplication automatically (e.g., 1111)

    # n: total length, k: block length, t: repetition count
    for n in range(2, max_digits + 1):
        for k in range(1, n):
            if n % k != 0:
                continue
            
            t = n // k
            if t < 2:
                continue

            start_x = pow10[k - 1]  # smallest k-digit number
            end_x = pow10[k] - 1    # largest k-digit number

            for x in range(start_x, end_x + 1):
                # Build the repeated number string-wise or math-wise
                # Math-wise to match Java logic:
                v = 0
                block_pow = pow10[k]
                for _ in range(t):
                    v = v * block_pow + x
                
                if v > max_val:
                    # For a fixed n, k, t, v increases with x
                    break
                
                results.add(v)

    return sorted(list(results))

def solve(input_file_path: Path) -> int:
    if not input_file_path.exists():
        raise FileNotFoundError(f"File not found: {input_file_path}")
        
    content = input_file_path.read_text(encoding='utf-8').strip()
    if not content:
        return 0

    # Parse ranges
    parts = content.split(",")
    ranges = []
    global_max = 0

    for token in parts:
        token = token.strip()
        if not token or '-' not in token:
            continue
        
        a_str, b_str = token.split('-', 1)
        a, b = int(a_str.strip()), int(b_str.strip())
        if a > b:
            a, b = b, a
        
        ranges.append((a, b))
        if b > global_max:
            global_max = b

    # Generate invalid IDs and create prefix sums
    invalid_ids = generate_all_repeated_numbers_up_to(global_max)
    
    # pref[i] is the sum of the first i elements
    pref = [0] * (len(invalid_ids) + 1)
    for i in range(len(invalid_ids)):
        pref[i + 1] = pref[i] + invalid_ids[i]

    total = 0
    for a, b in ranges:
        # lower_bound: first index i where invalid_ids[i] >= a
        left = bisect.bisect_left(invalid_ids, a)
        # upper_bound: first index i where invalid_ids[i] > b
        right = bisect.bisect_right(invalid_ids, b)
        
        if left < right:
            total += (pref[right] - pref[left])

    return total

def main():
    default_path = Path(r"E:\documents-container\advent-of-code\java-advent-of-code\inputs\2025-002\input.txt")
    input_path = Path(sys.argv[1]) if len(sys.argv) >= 2 else default_path
    
    try:
        ans = solve(input_path)
        print(ans)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()